File: apps/api/schemas/report.py
from datetime import date as Date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from schemas.inventory_snapshot import ProductInSnapshot


DeltaStatus = Literal["added", "consumed", "unchanged"]


class DeltaBoxItem(BaseModel):
    box_id: int
    box_code: str
    product_id: int
    quantity: int
    box_date: Date
    confidence_score: float | None = None


class DeltaItem(BaseModel):
    product: ProductInSnapshot
    previous_quantity: int
    current_quantity: int
    added_quantity: int
    removed_quantity: int
    delta: int
    status: DeltaStatus
    confidence_score: float | None = None
    added_boxes: list[DeltaBoxItem] = Field(default_factory=list)
    removed_boxes: list[DeltaBoxItem] = Field(default_factory=list)


class DailyDeltaResponse(BaseModel):
    date: Date
    current_snapshot_id: int
    baseline_date: Date
    baseline_snapshot_id: int | None = None
    items: list[DeltaItem]


class StockSummaryItem(BaseModel):
    product: ProductInSnapshot
    quantity: int


class StockSummaryResponse(BaseModel):
    snapshot_id: int
    snapshot_date: Date
    created_at: datetime
    items: list[StockSummaryItem]


class PlanVsActualItem(BaseModel):
    date: Date
    product: ProductInSnapshot
    planned_quantity: int
    actual_quantity: int
    variance: int
    snapshot_id: int | None = None


class PlanVsActualResponse(BaseModel):
    from_date: Date
    to_date: Date
    items: list[PlanVsActualItem]

File: apps/web/src/models/Report.ts
import { Product } from "@/models/Product";

export type DeltaStatus = "added" | "consumed" | "unchanged";

export interface DeltaBoxItem {
  box_id: number;
  box_code: string;
  product_id: number;
  quantity: number;
  box_date: string;
  confidence_score?: number | null;
}

export interface DailyDeltaItem {
  product: Product;
  previous_quantity: number;
  current_quantity: number;
  added_quantity: number;
  removed_quantity: number;
  delta: number;
  status: DeltaStatus;
  confidence_score?: number | null;
  added_boxes: DeltaBoxItem[];
  removed_boxes: DeltaBoxItem[];
}

export interface DailyDeltaResponse {
  date: string;
  current_snapshot_id: number;
  baseline_date: string;
  baseline_snapshot_id?: number | null;
  items: DailyDeltaItem[];
}

export interface StockSummaryItem {
  product: Product;
  quantity: number;
}

export interface StockSummaryResponse {
  snapshot_id: number;
  snapshot_date: string;
  created_at: string;
  items: StockSummaryItem[];
}

export interface PlanVsActualItem {
  date: string;
  product: Product;
  planned_quantity: number;
  actual_quantity: number;
  variance: number;
  snapshot_id?: number | null;
}

export interface PlanVsActualResponse {
  from_date: string;
  to_date: string;
  items: PlanVsActualItem[];
}
